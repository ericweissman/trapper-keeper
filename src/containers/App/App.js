import React, { Component } from 'react';
import { Switch, Route, withRouter } from 'react-router-dom'
import { connect } from 'react-redux';
import NotFound from '../../components/NotFound/NotFound'
import Header from '../../components/Header/Header'
import NoteForm from '../NoteForm/NoteForm'
import NoteArea from '../NoteArea/NoteArea'
import { fetchNotes } from '../../thunks/fetchNotes'
import PropTypes from 'prop-types';
import '../../Main.scss'
let shortID = require('short-id');


class App extends Component {

  componentDidMount = async () => {
    const url = 'http://localhost:3001/api/v1/notes'
    this.props.fetchNotes(url)
  }

  render() {
    return (
      <div className="App">
        <Header />
        <Switch>
          <Route exact path='/' component={NoteArea} />
          <Route path='/new-note' render={() => {
            const note = { title: '', id: shortID.generate(), timestamp: Date.now() }
            return <NoteForm note={note} items={[]} isEdit={false}/>
          }} />
          <Route path='/notes/:id' render={({ match }) => {
            const { id } = match.params
            const note = this.props.notes.find(note => note.id === id)
            const items = this.props.items.filter(item => item.noteID === id)
            if (note) {
              return <NoteForm note={note} items={items} isEdit={true}/>
            }
          }} />
          <Route component={NotFound} />
        </Switch>
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  notes: state.notes,
  items: state.items,
  isLoading: state.isLoading,
  error: state.error
})

export const mapDispatchToProps = (dispatch) => ({
  fetchNotes: (url) => dispatch(fetchNotes(url))
})

App.propTypes = {
  error: PropTypes.string,
  fetchNotes: PropTypes.func,
  isLoading: PropTypes.bool,
  items: PropTypes.array,
  notes: PropTypes.array,
}

App.defaultProps = {
  error: '',
  isLoading: true,
  items: [],
  notes: []
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App));