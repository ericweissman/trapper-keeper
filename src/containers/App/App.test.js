import React from 'react'
import { mapStateToProps, mapDispatchToProps } from './App'
import App from './App'
import { shallow } from 'enzyme'


describe('App', () => {
  let wrapper
  const mockProps = {
    notes: [{id: 1, title: 'Note Title'}],
    items: [],
    isLoading: true,
    error: '',
    fetchNotes: jest.fn()
  }
  it('should match the correct snapshot', () => {
    wrapper = shallow(<App {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  describe('mapStateToProps', () => {
    it('should return an object with notes, items, isLoading and error', () => {
      const mockState = {
        notes: [{}, {}],
        fakeState: [],
        fakeStateTwo: 2,
        items: [{}, {}, {}, {}],
        isLoading: true,
        error: ''
      }
      const expected = {
        notes: [{}, {}],
        items: [{}, {}, {}, {}],
        isLoading: true,
        error: ''
      }

      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('calls dispatch with fetchNotes when componentDidMount is called', () => {
      const mockDispatch = jest.fn()
      const mockURL = 'http://localhost:789423'
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.fetchNotes(mockURL)
      expect(mockDispatch).toHaveBeenCalled()
    })
  })
})