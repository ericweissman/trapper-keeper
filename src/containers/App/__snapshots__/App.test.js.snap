// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the correct snapshot 1`] = `
<Route>
  <Component />
</Route>
`;
